File: BSTIterator.cpp

#include<deque>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class BSTIterator {
    
public:

    deque<TreeNode*> nodes;

    BSTIterator(TreeNode* root) {
        iterateToNextLeftmostNodeFromCurrentRoot(root);
    }

    int next() {
        if (!hasNext()) {
            throw out_of_range("No Such Element Exception.");
        }

        TreeNode* nextNode = nodes.front();
        nodes.pop_front();
        iterateToNextLeftmostNodeFromCurrentRoot(nextNode->right);
        return nextNode->val;
    }

    bool hasNext() {
        return !nodes.empty();
    }

    void iterateToNextLeftmostNodeFromCurrentRoot(TreeNode* node) {
        while (node != nullptr) {
            nodes.push_front(node);
            node = node->left;
        }
    }
};

File: BSTIterator.js

function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

/**
 * @param {TreeNode} root
 */
var BSTIterator = function (root) {
    this.nodes = [];
    this.iterateToNextLeftmostNodeFromCurrentRoot(root);
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function () {
    if (!this.hasNext()) {
        throw "No Such Element Exception.";
    }

    let nextNode = this.nodes.pop();
    this.iterateToNextLeftmostNodeFromCurrentRoot(nextNode.right);
    return nextNode.val;
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function () {
    return this.nodes.length > 0;
};

/**
 * @param {TreeNode} node
 */
BSTIterator.prototype.iterateToNextLeftmostNodeFromCurrentRoot = function (node) {
    while (node !== null) {
        this.nodes.push(node);
        node = node.left;
    }
};
